src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpHandlerResourceV3Provider.cs

```diff
@@ -105,17 +116,44 @@ public CredentialPromptWebRequestHandler(IWebProxy proxy, ICredentials credentia
                         if (ProxyAuthenticationRequired(ex) &&
                             HttpHandlerResourceV3.PromptForProxyCredentials != null)
                         {
-                            // prompt use for proxy credentials.
-                            var credentials = await HttpHandlerResourceV3
-                                .PromptForProxyCredentials(request.RequestUri, Proxy, cancellationToken);
+                            ICredentials currentCredentials = Proxy.Credentials;
 
-                            if (credentials == null)
+                            try
                             {
-                                throw;
+                                await _credentialPromptLock.WaitAsync();
+
+                                // Check if the credentials have already changed
+                                if (beforeAuthId != _lastAuthId)
+                                {
+                                    continue;
+                                }
+
+                                // Limit the number of retries
+                                _authRetries++;
+                                if (_authRetries >= MaxAuthRetries)
```

Comment:
The scenarios seem very different, I don't see why they should be generalized and treated the same.

Examples:
* 501 status code
* Network timeout
* Retrying when a password typed in is incorrectly